README.md
# Fleet

[![Latest Version on Packagist](https://img.shields.io/packagist/v/aschmelyun/fleet.svg?style=flat-square)](https://packagist.org/packages/aschmelyun/fleet)
[![Total Downloads](https://img.shields.io/packagist/dt/aschmelyun/fleet.svg?style=flat-square)](https://packagist.org/packages/aschmelyun/fleet)

Easily run multiple Laravel Sail applications at the same time locally with custom domain names.

> Note: This is still new and may contain bugs, if you see something weird please [create an issue](https://github.com/aschmelyun/fleet/issues/new)

## Installation

You can install the package via composer:

```bash
composer require aschmelyun/fleet
```

## Getting Started

Navigate to the root directory of your Laravel application, and stop any Sail instance if it's running. Then, use the following command to add Fleet support to your app:

```bash
php artisan fleet:add
```

During setup you'll be prompted to enter in a domain name. Alternatively, you can pass it in through the command line:

```bash
php artisan fleet:add my-app.localhost
```

After the setup finishes, you can start up Sail just like normal:

```bash
./vendor/bin/sail up
```

And your site will be available at the domain you provided!

> Note: If you chose a domain that doesn't end in `.localhost`, you will need to add an entry to your hosts file to direct traffic to 127.0.0.1

## Additional Usage

By default, whenever you use `fleet:add`, a Docker network and container are both started to handle the traffic from your local domain name(s).

You can start this manually by using:

```bash
php artisan fleet:start
```

If you would like to remove Fleet support from an application and return it back to the default Docker setup, you can run:

```bash
php artisan fleet:remove
```

To stop and remove all Fleet containers and networks that are currently running on your system, you can use the command:

```bash
php artisan fleet:stop
```

## Why use this?

[Laravel Sail](https://laravel.com/docs/sail) uses Docker and Docker Compose to spin up containers that create a local development environment for your application.

By default, the containers are bound to the `:80` port of your local machine. Spinning up a second application results in a failure due to port conflicts, but you can always adjust it so that the second app is available at something like `:8081`.

This can have some unintended consequences though, and can get messy juggling communication between two different applications using port numbers.

Instead, Fleet provides a small set of commands that alter your `docker-compose.yml` file to provide support for [Traefik](https://hub.docker.com/_/traefik), a reverse proxy that runs on a Docker container.

When you add a site to Fleet, a network and a few labels are added to the Docker Compose entry for your application, and a main Traefik container is spun up to handle all local web traffic incoming to the `:80` port.

This configuration allows two or more Laravel Sail applications configured to different domains to resolve to their respective running containers.

For a more in-depth look at how this all ties together, check out [this video](https://www.youtube.com/watch?v=mZbLvGQqEvY) that I published on using Traefik with Docker Compose.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what's recently changed.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `aschmelyun/fleet` will be documented in this file.

## 1.0.2 - 2023-01-16

- Removes port bindings in `docker-compose.yml` for the app service
- Adds loadbalancer entry for traefik, bound to port 80
- Modifies the add command to use the first service name available, instead of assuming it's laravel.test

## 1.0.1 - 2023-01-15

- Adds `-a` flag when removing containers to search for all available ones
- Adds regex for filtering by exact name for fleet network and traefik container
- Fixes some capitalization

## 1.0.0 - 2023-01-15

- Initial release!


src/Commands/FleetAddCommand.php
<?php

namespace Aschmelyun\Fleet\Commands;

use Aschmelyun\Fleet\Fleet;
use Composer\InstalledVersions;
use Illuminate\Console\Command;
use Symfony\Component\Yaml\Yaml;

class FleetAddCommand extends Command
{
    public $signature = 'fleet:add {domain?}';

    public $description = 'Installs Fleet support onto the current application';

    public function handle(): int
    {
        // set the domain to the one the user provided, or ask what it should be
        $domain = $this->argument('domain');
        if (!$domain) {
            $domain = $this->ask('What domain name would you like to use for this app?', 'laravel.localhost');
        }

        $domainArray = explode('.', $domain);
        if (end($domainArray) !== 'localhost') {
            $this->line(' Just a quick note: You may need to update your hosts file to direct traffic to this domain. You can prevent this by using a domain that ends in `.localhost`');
        }

        // determine if laravel/sail is a required-dev package in the root composer file
        if (!InstalledVersions::isInstalled('laravel/sail')) {
            $this->error(' Laravel Sail is required for this package');
            $this->line(' For more information, check out https://laravel.com/docs/9.x/sail#installation');
        }

        // if the docker-compose.yml file isn't available, publish it
        if (!file_exists(base_path('docker-compose.yml')) && !file_exists(base_path('docker-compose.backup.yml'))) {
            $this->info('No docker-compose.yml file available, running sail:install...');
            $this->call('sail:install');
        }

        // copy the docker-compose.yml file for a backup
        if (file_exists(base_path('docker-compose.yml'))) {
            if ($this->confirm('This will modify your docker-compose.yml file, do you want to back it up first?')) {
                rename(base_path('docker-compose.yml'), base_path('docker-compose.backup.yml'));
            }
        }

        // determine what port 8081+ is available
        $port = 8081;
        while ($this->isPortTaken($port)) {
            $port++;
        }

        $file = base_path('.env');
        if (!file_exists($file)) {
            $this->error("Application .env file is missing, can't continue");

            return self::FAILURE;
        }

        $env = file_get_contents($file);
        $env = explode("\n", $env);

        $filteredEnvAppPort = array_filter($env, fn ($line) => str_starts_with($line, 'APP_PORT'));
        if (!empty($filteredEnvAppPort)) {
            $env[key($filteredEnvAppPort)] = "APP_PORT={$port}";
        } else {
            $insert = ["APP_PORT={$port}"];
            array_splice($env, 5, 0, $insert);
        }

        file_put_contents(base_path('.env'), implode("\n", $env));

        // add a modified docker-compose.yml file to include traefik labels
        $file = base_path('docker-compose.backup.yml');
        if (!file_exists($file)) {
            $file = base_path('docker-compose.yml');
        }

        if (!file_exists($file)) {
            $this->error('A docker-compose.yml file or a docker-compose.backup.yml file does not exist');

            return self::FAILURE;
        }

        $yaml = Yaml::parseFile($file);

        $heading = str_replace('.', '-', $domain);

        $service = $yaml['services'][array_keys($yaml['services'])[0]];
        unset($yaml['services'][array_keys($yaml['services'])[0]]);

        $yaml['services'] = [$heading => $service, ...$yaml['services']];

        $yaml['services'][$heading]['networks'][] = 'fleet';
        $yaml['services'][$heading]['labels'] = [
            "traefik.http.routers.{$heading}.rule=Host(`{$domain}`)",
            "traefik.http.services.{$heading}.loadbalancer.server.port=80",
        ];

        unset($yaml['services'][$heading]['ports'][0]);
        $yaml['services'][$heading]['ports'] = array_values($yaml['services'][$heading]['ports']);

        $yaml['networks']['fleet']['external'] = true;

        file_put_contents(base_path('docker-compose.yml'), Yaml::dump($yaml, 6));

        // call fleet:start to determine if the fleet network and traefik container is up
        $this->call('fleet:start');

        // return info back to the user
        $this->info(' ✨ All done! You can now run `./vendor/bin/sail up`');
        $this->newLine();

        return self::SUCCESS;
    }

    private function isPortTaken($port): bool
    {
        $process = Fleet::process("lsof -nP -iTCP:{$port} -sTCP:LISTEN");

        return boolval($process->getOutput());
    }
}
